scripts/scraper-ng/url-to-vfile.js

```diff
@@ -0,0 +1,49 @@
+const fetch = require("node-fetch");
+const isVFileMessage = require("is-vfile-message");
+const vfile = require("vfile");
+
+/**
+ * Returns a promise for the creation of a `VFile` object with the following
+ * details:
+ *
+ * - `pathname` is the URL pathname (e.g.,
+ *   `https://developer.mozilla.org/en-US/docs/Web/HTML/Element/span` becomes
+ *   `/en-US/docs/Web/HTML/Element/span`)
+ * - `contents` is the contents of the fetched page with query parameters `raw`
+ * - `data.url` is the original URL
+ *
+ */
+async function toVFile(url) {
+  const { pathname } = new URL(url);
+  const f = vfile({
+    path: pathname,
+    cwd: null,
+    data: { url: url }
+  });
+
+  try {
+    const response = await fetch(toRaw(url));
+    if (response.ok) {
+      f.contents = await response.text();
+    } else {
+      f.fail(`${response.status} ${response.statusText}`);
+    }
+  } catch (err) {
+    if (!isVFileMessage(err)) {
```

Comment:
This check seems a bit non-obvious. My first thought was, "in what conditions will `try` throw without there being a fatal error"? But I think what's actually happening is that `fail()` is setting `fatal`. So really the error-handling code is still duplicated, except that one of the copies is hidden inside `fail()`. It would seem clearer to just throw in this case too, then there's only one place it gets handled. But it's a super-minor point, and probaby subjective anyway.

Reply:
I wasn't quite sure what you were getting at here but I think I rubber ducked my way to it (I've left my notes about this below, to show what I was thinking). I guess what should happen here is that I have a wrapper for `fetch` that either re-throws `fetch`'s exceptions or creates a new exception and then have the `vfile` error handling happen in only one place.

I'll push some changes for this shortly.

---

I agree it's a bit hard to follow but I'm not quite sure how to make it more obvious. Here's what I'm trying to capture with the existing code:

* The successful path: `fetch` succeeds at the network level and the response is successful (i.e., the page loads).
* An error response: `fetch` succeeds at the network level but the response is an error (e.g., the server responds with a 404). This corresponds to the `try` block's `else` condition. In the existing code, `f.fail()` throws and is swallowed by the `catch` block.
* Fetch failure: `fetch` itself throws (i.e., it can't make a connection to the server). In the existing code, a new error is created.

In the error cases, I want the `vfile` to have one and only one fatal error attached to it and then return the vfile (i.e., I never want to actually throw from `toVFile` at all).